src/protocol/context/malicious.rs

```diff
@@ -279,52 +257,151 @@ impl<'a, F: Field> ContextInner<'a, F> {
         Ok(m)
     }
 
-    async fn upgrade(
+    async fn upgrade<T, M>(&self, input: T) -> Result<M, Error>
+    where
+        for<'u> UpgradeContext<'u, F>: UpgradeToMalicious<T, M>,
+    {
+        UpgradeContext {
+            upgrade_ctx: self.upgrade_ctx.clone(),
+            inner: self,
+            record_binding: NoRecord,
+        }
+        .upgrade(input)
+        .await
+    }
+
+    async fn upgrade_with_sparse<SS: Substep>(
         &self,
-        record_id: RecordId,
-        x: Replicated<F>,
+        step: &SS,
+        input: Replicated<F>,
         zeros_at: ZeroPositions,
     ) -> Result<MaliciousReplicated<F>, Error> {
-        self.upgrade_one(self.upgrade_ctx.clone(), record_id, x, zeros_at)
-            .await
+        UpgradeContext {
+            upgrade_ctx: self.upgrade_ctx.narrow(step),
+            inner: self,
+            record_binding: NoRecord,
+        }
+        .upgrade_sparse(input, zeros_at)
+        .await
     }
 
-    async fn upgrade_with<SS: Substep>(
+    async fn upgrade_for_record_with<SS: Substep, T, M>(
         &self,
         step: &SS,
         record_id: RecordId,
-        x: Replicated<F>,
+        input: T,
+    ) -> Result<M, Error>
+    where
+        for<'u> UpgradeContext<'u, F, RecordId>: UpgradeToMalicious<T, M>,
+    {
+        UpgradeContext {
+            upgrade_ctx: self.upgrade_ctx.narrow(step),
+            inner: self,
+            record_binding: record_id,
+        }
+        .upgrade(input)
+        .await
+    }
+}
+
+/// Helper to prevent using the record ID multiple times to implement an upgrade.
```

Comment:
I think that this code from here on down could be moved to a new file (upgrade.rs).

Reply:
Sorry, I missed these comments before merging the upgrades PR. I agree it makes sense to move to a separate file, but I think I'll leave it alone for right now since it's no longer part of this PR.